crates/edr_napi/src/trace/message_trace.rs

```diff
@@ -58,10 +65,9 @@ pub struct CallMessageTrace {
     pub depth: u32,
     // `BaseEvmMessageTrace`
     pub code: Uint8Array,
-    pub steps: Vec<Either4<EvmStep, PrecompileMessageTrace, CreateMessageTrace, CallMessageTrace>>,
-    // TODO: Will be later filled on the JS side but we should port to ContractsIdentifier in Rust
-    // This is explicitly `any` on the JS side to side-step the type-checking until we port
-    pub bytecode: Option<Value>,
+    pub steps: Vec<Either4<EvmStep, PrecompileMessageTrace, CallMessageTrace, CreateMessageTrace>>,
+    #[napi(ts_type = "Bytecode")]
```

Comment:
If this is now filled on the Rust side, is there a reason we can't define an N-API `Bytecode` type?

Reply:
We do define and use `Bytecode` elsewhere already but there's some subtle nuance here:

1. We can't use `ClassInstance` here because it internally holds a `&'static mut T` so that's UB because we probably will often have traces that point to the same bytecode
2. I introduced the `ClassInstanceRef` that guards the access with a `RefCell`, however it's not possible to directly use it here as it'd require `ToNapiValue` and `FromNapiValue` implementation...
3. and while it might be feasible to implement it using `ToNapiValue` (it can pass the underlying object), however...
4. implementing `FromNapiValue` would be very tricky:
  a. the instances are shared, so on the Rust side are wrapped in `Rc` and internally uses `napi::Ref` for napi-side ref-counting of the underlying object, so...
  b. converting at the FFI boundary without a side-table would have to create new Rc'd objects; ignoring the potential perf drawbacks, it would also be incorrect because it would lead to multiple `RefCells` guarding the same memory access and...
  c. introducing the side-table means introducing a static map per JS class on the Rust side, hooking into the napi lifecycle to manage the objects on the Rust side once they're freed on the JS side, modifying the `ClassInstanceRef` helper to do the same on the Rust side...
  d. ...at which point we'd be reimplementing quite a bit of higher-level convenience code that arguably should be on the napi-rs side itself; hopefully, we won't need as much JS glue anymore once the port is complete.
  
Let me know if the context makes sense; I'm more than happy if we simplified this somehow but I didn't manage to find a way to make it both safe and convenient :cry:

Reply:
It sounds very complicated. I'll trust your judgement!

Reply:
Thank you! I'll try to embed a bit more context and/or more explain these things more concisely next time :pray: